File: javascript/235-lowest-common-ancestor-of-a-binary-search-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */

/**
 * @param {TreeNode} root
 * @param {TreeNode} p
 * @param {TreeNode} q
 * @return {TreeNode}
 */
 var lowestCommonAncestor = function(root, p, q) {
    while(root !== null) {
        if (root.val < p.val && root.val < q.val ) {
            root = root.right
        } else if (root.val > p.val && root.val > q.val) {
            root = root.left
        } else {
            return root
        }
    }
};

File: ruby/235-Lowest-Common-Ancestor-of-a-Binary-Search-Tree.rb
def lowest_common_ancestor(root, p, q)
  return root if root.val == p.val || root.val == q.val
  if root.val.between?(p.val, q.val) || root.val.between?(q.val, p.val)
    return root
  end

  if root.val > p.val
    lowest_common_ancestor(root.left, p, q)
  else
    lowest_common_ancestor(root.right, p, q)
  end
end
